Added GraphSAGE attention aggregator

adocherty opened the pull request:
Added n GAT-style attention aggregator for GraphSAGE. This performs the same attention computation as GAT on a batch of input nodes. Dropout is handled externally in the same way as the default GraphSAGE code (different to GAT which would require dropout within the aggregator).

Currently the aggregator is single-headed only. We can extend this to multi-headed in a more general way by adding the multi-head support in the GraphSAGE model class, enabling multiple-head models with different aggregators than just the attention aggregator.

Additional changes are to build a more flexible MLP model when there are zero samples in all aggregators (required as the attention aggregator wasn't compatible with how I did this previously).

youph commented:
@adocherty before merging, could you please update README to reflect the fact that GraphSAGE now has more than one aggregator?

adocherty commented:
@youph I've added a quick note in the README stating we have multiple aggregators.  I've also updated the docstring for GraphSAGE to include this information as well.
